doc/manual/compiler.rst

```diff
@@ -54,74 +72,80 @@ The Python types correspond to ARTIQ type annotations as follows:
-interpreter.
+User-defined classes are supported, provided their attributes are of other supported types (attributes that are not used in the kernel are ignored and thus unrestricted). When several instances of a user-defined class are referenced from the same kernel, every attribute must have the same type in every instance of the class.
 
-Empty lists do not have valid list element types, so they cannot be used in the kernel.
+Basic ARTIQ Python
+^^^^^^^^^^^^^^^^^^
 
-In kernels, lifetime of allocated values (e.g. lists or numpy arrays) might not be correctly
-tracked across function calls (see `#1497 <https://github.com/m-labs/artiq/issues/1497>`_,
-`#1677 <https://github.com/m-labs/artiq/issues/1677>`_) like in this example ::
+Basic Python features can broadly be used inside kernels without further compunctions. This includes loops (``for`` / ``while`` / ``break`` / ``continue``), conditionals (``if``, ``else``, ``elif``), functions, exceptions, ``try`` / ``except`` / ``else`` blocks,  and statically typed variables of any supported types.
 
-    @kernel
-    def func(a):
-        return a
+Kernel code can call host functions without any additional ceremony. However, such functions are assumed to return ``None``, and if a value other than ``None`` is returned, an exception is raised. To call a host function returning a value other than ``None`` its return type must be annotated, using the standard Python syntax, e.g.: ::
 
-    class ProblemReturn1(EnvExperiment):
-        def build(self):
-            self.setattr_device("core")
+    def return_four() -> TInt32:
+        return 4
 
-        @kernel
-        def run(self):
-            # results in memory corruption
-            return func([1, 2, 3])
+Kernels can freely modify attributes of objects shared with the host. However, by necessity, these modifications are actually applied to local copies of the objects, as the latency of immediate writeback would be unsupportable in a real-time environment. Instead, modifications are written back *when the kernel completes;* notably, this means RPCs called by a kernel itself will only have access to the unmodified host version of the object, as the kernel hasn't finished execution yet. In some cases, accessing data on the host is better handled by calling RPCs specifically to make the desired modifications. 
+
+.. warning::
 
-This results in memory corruption at runtime.
+    Kernels *cannot and should not* return lists, arrays, or strings they have created, or any objects containing them; in the absence of a heap, the way these values are allocated means they cannot outlive the kernels they are created in. Trying to do so will normally be discovered by lifetime tracking and result in compilation errors, but in certain cases lifetime tracking will fail to detect a problem and experiments will encounter memory corruption at runtime. For example: :: 
+
+        def func(a):
```

Comment:
I know it was in the original, but should we still have negative code examples in the manual? Could be confusing if skimming through the manual, and if it's in a Github issue, linking to it would probably be the better option.

Does remind me that Rust kind of works around negative examples with a special icon indicating that the code is broken/will fail to compile.

Reply:
Nevermind, disregard this - it's in a warning section, very visible as a potential problem.